File: backend/app/services/storage_service.py
from typing import Dict, List
from ..models.medicine import Medicine

class StorageService:
    # Standard storage conditions and their requirements
    STORAGE_CONDITIONS = {
        "room_temperature": {
            "temp_min": 15,
            "temp_max": 25,
            "humidity_max": 60,
            "description": "Store at room temperature (15-25°C)",
            "recommendations": [
                "Keep in a cool, dry place",
                "Avoid direct sunlight",
                "Keep away from heat sources"
            ]
        },
        "refrigerated": {
            "temp_min": 2,
            "temp_max": 8,
            "humidity_max": 65,
            "description": "Store in refrigerator (2-8°C)",
            "recommendations": [
                "Keep refrigerated",
                "Do not freeze",
                "Keep in original packaging"
            ]
        },
        "cool": {
            "temp_min": 8,
            "temp_max": 15,
            "humidity_max": 65,
            "description": "Store in a cool place (8-15°C)",
            "recommendations": [
                "Keep in a cool place",
                "Protect from light",
                "Keep container tightly closed"
            ]
        }
    }

    @staticmethod
    def get_storage_recommendations(medicine: Medicine) -> Dict:
        """Get storage recommendations for a medicine"""
        try:
            form = medicine.form.lower()
            storage_type = "room_temperature"  # default

            # Determine storage type based on medicine form
            if form in ["liquid", "suspension", "syrup"]:
                storage_type = "cool"
            elif form in ["injection", "vaccine"]:
                storage_type = "refrigerated"

            # Get standard conditions
            conditions = StorageService.STORAGE_CONDITIONS[storage_type]

            # Add form-specific recommendations
            form_recommendations = StorageService.get_form_specific_recommendations(form)
            
            return {
                "storage_type": storage_type,
                "conditions": conditions,
                "form_specific_recommendations": form_recommendations
            }

        except Exception as e:
            print(f"Error getting storage recommendations: {str(e)}")
            return None

    @staticmethod
    def get_form_specific_recommendations(form: str) -> List[str]:
        """Get recommendations specific to medicine form"""
        recommendations = {
            "tablet": [
                "Keep in original container",
                "Protect from moisture",
                "Keep container tightly closed"
            ],
            "capsule": [
                "Keep in original container",
                "Protect from moisture",
                "Store in a dry place"
            ],
            "liquid": [
                "Do not freeze",
                "Shake well before use",
                "Keep bottle tightly closed"
            ],
            "cream": [
                "Do not refrigerate unless specified",
                "Keep tube properly closed",
                "Store at room temperature"
            ],
            "injection": [
                "Keep in original packaging",
                "Protect from light",
                "Check expiration date before use"
            ]
        }
        
        return recommendations.get(form.lower(), [
            "Keep in original container",
            "Follow package instructions",
            "Store safely away from children"
        ]) 

File: lib/services/storage_service.dart
import '../models/medicine.dart';
import 'api_service.dart';

class StorageService {
  static Future<Map<String, dynamic>> getMedicineStorageRecommendations(
    String medicineId,
  ) async {
    try {
      final response = await ApiService.get(
        '/medicines/$medicineId/storage-recommendations',
      );
      return response;
    } catch (e) {
      throw Exception('Failed to get storage recommendations: $e');
    }
  }

  // Helper methods for displaying storage information
  static String getStorageConditionsText(Map<String, dynamic> conditions) {
    return conditions['description'] ?? 'Store according to package instructions';
  }

  static List<String> getAllRecommendations(Map<String, dynamic> storageData) {
    List<String> allRecommendations = [];
    
    // Add general storage conditions recommendations
    if (storageData['conditions'] != null && 
        storageData['conditions']['recommendations'] != null) {
      allRecommendations.addAll(
        List<String>.from(storageData['conditions']['recommendations'])
      );
    }
    
    // Add form-specific recommendations
    if (storageData['form_specific_recommendations'] != null) {
      allRecommendations.addAll(
        List<String>.from(storageData['form_specific_recommendations'])
      );
    }
    
    return allRecommendations;
  }

  static bool isRefrigerated(Map<String, dynamic> storageData) {
    return storageData['storage_type'] == 'refrigerated';
  }

  static String getTemperatureRange(Map<String, dynamic> conditions) {
    final tempMin = conditions['temp_min'];
    final tempMax = conditions['temp_max'];
    return '$tempMin-$tempMax°C';
  }
} 